loader/src/main/java/net/neoforged/fml/common/asm/enumextension/EnumProxy.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.fml.common.asm.enumextension;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import org.jetbrains.annotations.ApiStatus;
+
+public final class EnumProxy<T extends Enum<T> & IExtensibleEnum> {
+    private final Class<T> enumClass;
+    private final List<Object> parameters;
+    private T enumValue;
+
+    public EnumProxy(Class<T> enumClass, Object... parameters) {
+        this(enumClass, Arrays.asList(parameters));
+    }
+
+    public EnumProxy(Class<T> enumClass, List<Object> parameters) {
+        this.enumClass = enumClass;
+        this.parameters = parameters;
+    }
+
+    @ApiStatus.Internal
+    public Object getParameter(int idx) {
+        return parameters.get(idx);
+    }
+
+    public T getValue() {
+        if (enumValue == null) {
+            try {
+                var loader = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE).getCallerClass().getClassLoader();
+                Class.forName(enumClass.getName(), true, loader);
+            } catch (ClassNotFoundException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return Objects.requireNonNull(enumValue, "Enum not initialized");
+    }
+
+    @ApiStatus.Internal
+    public void setValue(T value) {
+        this.enumValue = value;
```

Comment:
There's a small chance that this could run into concurrency issues. Though I'm not sure if locking in the get method is a good idea as it's a relatively hot path.

Reply:
Shouldn't this be prevented by the explicit classload in the getter? I thought there was some kind of lock involved to prevent a class from being initialized on multiple threads.

Reply:
> I thought there was some kind of lock involved to prevent a class from being initialized on multiple threads.

That is correct; for our purposes, `<clinit>` behaves as though synchronized to the class type. It'd probably be wise to mark `enumValue` as `volatile` though.